src/Microsoft.Identity.Client/Telemetry.cs

```diff
@@ -42,11 +44,142 @@ public void RegisterReceiver(Receiver r)
+            {
+                // if this is the first event associated to this
+                // RequestId we need to initialize a new List to hold
+                // all of sibling events
+                List<EventBase> events = new List<EventBase>();
+                events.Add(eventToStop);
+                CompletedEvents[requestId] = events;
+            }
+            else
+            {
+                // if this event shares a RequestId with other events
+                // just add it to the List
+                CompletedEvents[requestId].Add(eventToStop);
+            }
+
+            // Mark this event as no longer in progress
+            EventsInProgress.Remove(eventKey);
+        }
+
+        internal void Flush(string requestId)
+        {
+            if (_receiver == null)
+            {
+                return;
+            }
+
+            // check for orphaned events...
+            List<EventBase> orphanedEvents = CollateOrphanedEvents(requestId);
+            // Add the OrphanedEvents to the completed EventList
+            if (!CompletedEvents.ContainsKey(requestId))
+            {
+                // No completed Events returned for RequestId
+                return;
+            }
+
+            CompletedEvents[requestId].AddRange(orphanedEvents);
+
+            List<EventBase> eventsToFlush = CompletedEvents[requestId];
+            CompletedEvents.Remove(requestId);
+
+            if (TelemetryOnFailureOnly)
+            {
+                // iterate over Events, if the ApiEvent was successful, don't dispatch
+                bool shouldRemoveEvents = false;
+
+                foreach (var anEvent in eventsToFlush)
+                {
+                    var apiEvent = anEvent as ApiEvent;
+                    if (apiEvent != null)
+                    {
+                        shouldRemoveEvents = apiEvent.WasSuccessful;
+                        break;
+                    }
+                }
+
+                if (shouldRemoveEvents)
+                {
+                    eventsToFlush.Clear();
+                }
+            }
+
+            if (eventsToFlush.Count > 0)
+            {
+                eventsToFlush.Insert(0, new DefaultEvent(ClientId));
+                _receiver(eventsToFlush.Cast<Dictionary<string, string>>().ToList());
+            }
+        }
+
+        private List<EventBase> CollateOrphanedEvents(String requestId)
+        {
+            var orphanedEvents = new List<EventBase>();
+            foreach (var key in EventsInProgress.Keys)
+            {
+                if (key.Item1 == requestId)
+                {
+                    // The orphaned event already contains its own start time, we simply collect it
+                    orphanedEvents.Add(EventsInProgress[key]);  
+                    EventsInProgress.Remove(key);
+                }
+            }
+            return orphanedEvents;
+        }
+
+        public string ClientId { get; set; } = "undefined";
```

Comment:
True, and we did implement in the way your suggest: the app instances is already calling this setter to set a proper ClientId. The motivation for an `"undefined"` here is to make the default value to be a string, rather than an implicit `null`. You know, just in case. That being said, it shouldn't make a difference, providing that our App instances implementation is correct. So I can remove that in next commit.